File: (C++)ProblemSolving/ProblemSolving/[BOJ_1922]네트워크 연결.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

struct Node {
	int cost;
	int u, v;
	bool operator<(const Node& rhs) {
		return cost < rhs.cost;
	}
};
int parent[1001];

int find(int u) {
	if (parent[u] == 0) return u;
	else return find(parent[u]);
}
bool merge(int u, int v) {
	u = find(u);
	v = find(v);
	if (u == v) return false;;
	parent[u] = v;
	return true;
}


int main(void) {
	int n, m;
	cin >> n >> m;
	vector<Node> edge;
	for (int i = 0; i < m; i++) {
		int u, v, c;
		cin >> u >> v >> c;
		edge.push_back({ c, u, v });
	}
	sort(edge.begin(), edge.end());
	int answer = 0;
	for (int i = 0; i < edge.size(); i++) {
		if (merge(edge[i].u, edge[i].v)) {
			answer += edge[i].cost;
		}
	}
	cout << answer;
	return 0;
}

File: (Pyton)ProblemSolving/ProblemSovling/그래프/MST/[BOJ_1922]네트워크 연결.py
N = int(input())
M = int(input())
edges = []
for _ in range(M):
    u, v, c = list(map(int, input().split()))
    edges.append([u, v, c])

parent = [-1 for _ in range(N+1)]


def find(u):
    if parent[u] == -1:
        return u
    else:
        parent[u] = find(parent[u])
        return parent[u]


def merge(u, v):
    u = find(u)
    v = find(v)

    if u == v:
        return -1

    parent[u] = v


answer = 0
edges.sort(key=lambda x: x[2])
for edge in edges:
    u, v, c = edge
    if merge(u, v) != -1:
        answer += c

print(answer)
